<a href="https://colab.research.google.com/github/Sana-0511/Document_Summarization-/blob/main/Transformer_BERT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install transformers
!pip install bert-extractive-summarizer

In [ ]:
# Install the required utility
!apt-get install wget

# Import the necessary libraries
import zipfile
import os

# Provide the Google Drive link
google_drive_link = 'https://drive.google.com/uc?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI'

# Specify the output file name
zip_file = 'example.zip'

# Download the zip file
!wget --no-check-certificate $google_drive_link -O $zip_file

# Create a directory to extract the contents
extracted_dir = 'extracted_content'
os.makedirs(extracted_dir, exist_ok=True)

# Extract the contents of the zip file
with zipfile.ZipFile(zip_file, 'r') as zip_ref:
    zip_ref.extractall(extracted_dir)

# List the files in the extracted directory
extracted_files = os.listdir(extracted_dir)
print("Files extracted:", extracted_files)

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
wget is already the newest version (1.21.2-2ubuntu1).
0 upgraded, 0 newly installed, 0 to remove and 24 not upgraded.
--2024-01-13 08:45:37--  https://drive.google.com/uc?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI
Resolving drive.google.com (drive.google.com)... 173.194.213.101, 173.194.213.113, 173.194.213.100, ...
Connecting to drive.google.com (drive.google.com)|173.194.213.101|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://drive.usercontent.google.com/download?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI [following]
--2024-01-13 08:45:37--  https://drive.usercontent.google.com/download?id=1FV5aYhpvl6NFrH7aYzEVg8pmO_yokYoI
Resolving drive.usercontent.google.com (drive.usercontent.google.com)... 173.194.218.132, 2607:f8b0:400c:c14::84
Connecting to drive.usercontent.google.com (drive.usercontent.google.com)|173.194.218.132|:443... connected.
HTTP request sent

In [ ]:
# Import necessary modules from the transformers library
from transformers import pipeline, BertTokenizer, BertForSequenceClassification

# Function to read the content of a file
def read_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        # Read the entire content of the file
        text = file.read()
    return text

In [ ]:
from transformers import BertTokenizer, BertModel
from summarizer import Summarizer

def bert_extractive_summarization(text, num_sentences=3):
    # Load pre-trained BERT tokenizer
    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

    # Tokenize the input text using BERT tokenizer
    inputs = tokenizer(text, return_tensors='pt', max_length=512, truncation=True)

    # Load pre-trained BERT model for extractive summarization
    model = Summarizer()

    # Get summary
    summary = model(text, max_length=num_sentences*50)  # Adjust max_length as needed

    return summary

In [ ]:
# Main execution block
if __name__ == "__main__":
    # Specify the path to the text file
    file_path = '/content/extracted_content/example/msft.txt'

    # Read the content of the file
    text_content = read_file(file_path)

    # Set the number of sentences desired in the summary
    num_summary_sentences = 5

    # Generate the summary using BERT extractive summarization
    summary = bert_extractive_summarization(text_content, num_summary_sentences)

    # Print the generated summary
    print("Summary:\n", summary)


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Summary:
 In an attempt to build an AI-ready workforce, Microsoft announced Intelligent Cloud Hub which has been launched to empower the next generation of students with AI-ready skills. The program was developed to provide job ready skills to programmers who wanted to hone their skills in AI and data science with a series of online courses which featured hands-on labs and expert instructors as well.


In an attempt to build an AI-ready workforce, Microsoft announced Intelligent Cloud Hub which has been launched to empower the next generation of students with AI-ready skills
Envisioned as a three-year collaborative program, Intelligent Cloud Hub will support around 100 institutions with AI infrastructure, course content and curriculum, developer support, development tools and give students access to cloud and AI services
As part of the program, the Redmond giant which wants to expand its reach and is planning to build a strong developer ecosystem in India with the program will set up the core AI infrastructure and IoT Hub for the selected campuses
The company will provide AI development tools and Azure AI services such as Microsoft Cognitive Services, Bot Services and Azure Machine Learning.According to Manish Prakash, Country General Manager-PS, Health and Education, Microsoft India, said, "With AI being the defining technology of our time, it is transforming lives and industry and the jobs of tomorrow will require a different skillset
This will require more collaborations and training and working with AI
That’s why it has become more critical than ever for educational institutions to integrate new cloud and AI technologies
The program is an attempt to ramp up the institutional set-up and build capabilities among the educators to educate the workforce of tomorrow." The program aims to build up the cognitive skills and in-depth understanding of developing intelligent cloud connected solutions for applications across industry
Earlier in April this year, the company announced Microsoft Professional Program In AI as a learning track open to the public
The program was developed to provide job ready skills to programmers who wanted to hone their skills in AI and data science with a series of online courses which featured hands-on labs and expert instructors as well
This program also included developer-focused AI school that provided a bunch of assets to help build AI skills.

Summary:
The company will provide AI development tools and Azure AI services such as Microsoft Cognitive Services, Bot Services and Azure Machine Learning. Earlier in April this year, the company announced Microsoft Professional Program In AI as a learning track open to the public.

In [ ]:
!pip install rouge
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.287292817679558, 'p': 1.0, 'f': 0.44635192786310307}, 'rouge-2': {'r': 0.2108843537414966, 'p': 0.9841269841269841, 'f': 0.3473389326676553}, 'rouge-l': {'r': 0.287292817679558, 'p': 1.0, 'f': 0.44635192786310307}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.287
F1-score (f): 0.446

ROUGE-2:

Precision (p): 0.984
Recall (r): 0.211
F1-score (f): 0.347

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.287
F1-score (f): 0.446

These scores indicate high precision, suggesting that the generated summary contains a significant amount of correct unigrams (ROUGE-1) and overlaps well in terms of the longest common subsequence (ROUGE-L). However, the recall values are lower, indicating that the generated summary may not capture all the important content from the original text.

In [ ]:
# Main execution block
if __name__ == "__main__":
    # Specify the path to the text file
    file_path = '/content/extracted_content/example/computer.txt'

    # Read the content of the file
    text_content = read_file(file_path)

    # Set the number of sentences desired in the summary
    num_summary_sentences = 5

    # Generate the summary using BERT extractive summarization
    summary = bert_extractive_summarization(text_content, num_summary_sentences)

    # Print the generated summary
    print("Summary:\n", summary)


Summary:
 A computer is a machine that can be programmed to carry out sequences of arithmetic or logical operations automatically. These programs enable computers to perform a wide range of tasks.


/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


A computer is a machine that can be programmed to carry out sequences of arithmetic or logical operations automatically. Modern digital electronic computers can perform generic sets of operations known as programs. These programs enable computers to perform a wide range of tasks. A computer system is a nominally complete computer that includes the hardware, operating system, and peripheral equipment needed and used for full operation. This term may also refer to a group of computers that are linked and function together, such as a computer network or computer cluster. A broad range of industrial and consumer products use computers as control systems. Simple special-purpose devices like microwave ovens and remote controls are included, as are factory devices like industrial robots and computer-aided design, as well as general-purpose devices like personal computers and mobile devices like smartphones. Computers power the Internet, which links billions of other computers and users.

Summary: A computer is a machine that can be programmed to carry out sequences of arithmetic or logical operations automatically. These programs enable computers to perform a wide range of tasks.

In [ ]:
!pip install rouge
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.28125, 'p': 1.0, 'f': 0.43902438681737066}, 'rouge-2': {'r': 0.19858156028368795, 'p': 0.9655172413793104, 'f': 0.32941176187612464}, 'rouge-l': {'r': 0.28125, 'p': 1.0, 'f': 0.43902438681737066}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.281
F1-score (f): 0.439

ROUGE-2:

Precision (p): 0.966
Recall (r): 0.199
F1-score (f): 0.329

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.281
F1-score (f): 0.439

These scores indicate high precision, suggesting that the generated summary contains a significant amount of correct unigrams (ROUGE-1) and overlaps well in terms of the longest common subsequence (ROUGE-L). However, the recall values are lower, indicating that the generated summary may not capture all the important content from the original text.

In [ ]:
# Main execution block
if __name__ == "__main__":
    # Specify the path to the text file
    file_path = '/content/extracted_content/example/space.txt'

    # Read the content of the file
    text_content = read_file(file_path)

    # Set the number of sentences desired in the summary
    num_summary_sentences = 5

    # Generate the summary using BERT extractive summarization
    summary = bert_extractive_summarization(text_content, num_summary_sentences)

    # Print the generated summary
    print("Summary:\n", summary)


Summary:
 A white hole is a bizarre cosmic object which is intensely bright, and from which matter gushes rather than disappears. But then, quantum effects occurring around the surface of the black hole halt further collapse to a singularity, and instead begin to gradually turn the black hole into a white hole that’s spewing out the original star matter again.


/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


A white hole is a bizarre cosmic object which is intensely bright, and from which matter gushes rather than disappears. In other words, it’s the exact opposite of a black hole. But unlike black holes, there’s no consensus about whether white holes exist, or how they’d be formed.They are predicted by Einstein’s theory of gravity, and are most often mentioned in the context of ‘wormholes’, in which a black hole acts as the entry point to a tunnel through space and time, ending in a white hole somewhere else in the Universe. But this is deeply controversial, because Einstein’s theory predicts the existence of a so-called singularity at the centre of black holes – a state of infinite gravity which would prevent anything from passing through to the white hole on the other side. The process starts when an old massive star collapses under its own weight and forms a black hole. But then, quantum effects occurring around the surface of the black hole halt further collapse to a singularity, and instead begin to gradually turn the black hole into a white hole that’s spewing out the original star matter again. The process is mind-bendingly slow, though, so we may be in for a very long wait to find out if white holes really exist.

Summary: A white hole is a bizarre cosmic object which is intensely bright, and from which matter gushes rather than disappears. In other words, it’s the exact opposite of a black hole.

In [ ]:
!pip install rouge
from rouge import Rouge

# Your original text and generated summary
original_text = text_content
generated_summary = summary

# Initialize the Rouge object
rouge = Rouge()

# Calculate ROUGE scores
scores = rouge.get_scores(generated_summary, original_text)

# Print the scores
print("ROUGE Scores:")
print(scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.3235294117647059, 'p': 1.0, 'f': 0.4888888851950618}, 'rouge-2': {'r': 0.27411167512690354, 'p': 0.9818181818181818, 'f': 0.4285714251590451}, 'rouge-l': {'r': 0.3235294117647059, 'p': 1.0, 'f': 0.4888888851950618}}]


ROUGE-1:

Precision (p): 1.000
Recall (r): 0.324
F1-score (f): 0.489

ROUGE-2:

Precision (p): 0.982
Recall (r): 0.274
F1-score (f): 0.429

ROUGE-L:

Precision (p): 1.000
Recall (r): 0.324
F1-score (f): 0.489

These scores suggest high precision, indicating that the generated summary contains a substantial amount of correct unigrams (ROUGE-1) and overlaps well in terms of the longest common subsequence (ROUGE-L). However, the recall values are lower, indicating that the generated summary may not capture all the important content from the original text.

## REFERENCES
https://towardsdatascience.com/extractive-summarization-using-bert-966e912f4142

https://iq.opengenus.org/bert-for-text-summarization/
